fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Utils.java

```diff
@@ -827,4 +832,64 @@ public static OptExpression mergeProjection(OptExpression input,
         }
         return input;
     }
+
+    /**
+     * Check if the operator has applied the rule
+     * @param op input operator to be checked
+     * @param ruleMask specific rule mask
+     * @return true if the operator has applied the rule, false otherwise
+     */
+    public static boolean isOpAppliedRule(Operator op, int ruleMask) {
```

Comment:
int only has 32-bit, is there a risk of mask bits overflowing?
How about use BigInteger or RoaringBitmap instead?

Reply:
Yes. you are right. 

For now it's enough yet. we can do it later?